### PR TITLE
APID-471 - Scheduled sending of emails

### PR DESCRIPTION
**Pull request by worthydolt**

### CODE REVIEW

File: app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class SentEmailService@Inject()(emailConnector: GatekeeperEmailConnector,
+                                draftEmailService: DraftEmailService,
+                                sentEmailRepository: SentEmailRepository)
+                                (implicit val ec: ExecutionContext) extends Logging {
+
+   def sendAllPendingEmails: Future[Int] = {
```
Comment (ghchapman) [inline app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala:34]: Is this name misleading? As far as I can tell it just sends the next email, not all pending.

File: app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     def updateEmailStatusToSent(email:SentEmail): Future[SentEmail] = {
+       sentEmailRepository.markSent(email)
+     }
+
+     def handleEmailSendingFailed(email:SentEmail): Future[SentEmail] = {
+       logger.info(s"Handling failed message for email with failedCount of ${email.failedCount}")
+       if (email.failedCount > 2)  {
+         sentEmailRepository.markFailed(email)
+       } else {
+         logger.info(s"Incrementing failed counter for email ${email}")
```
Comment (ghchapman) [inline app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala:45]: Does logging the whole email here include PII (ie recipient email address)?

File: app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+       } yield email
+     }
+
+      def findNextEmail: Future[Option[SentEmail]] = {
+       sentEmailRepository.findNextEmailToSend
+     }
+
+     def sendEmail (email: DraftEmail): Future[Int] = {
+       logger.info(s"Retrieved email wth id ${email.emailUUID}")
+       val emailRequestedData = SendEmailRequest(email.recipients, email.templateData.templateId, email.templateData.parameters,
```
Comment (ghchapman) [inline app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala:70]: This looks like it's sending to all the recipients on the draft email, whereas my understanding was each invocation of the scheduled task would just send to one recipient, ie the one on the SentEmail, to drip feed the email service.

**Review by srinivasalumunagala — APPROVED** [review]

File: app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala (hunk 1 above)
Comment (worthydolt) [reply app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala]: Yep, quite so. Renamed

File: app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala (hunk 2 above)
Comment (worthydolt) [reply app/uk/gov/hmrc/gatekeeperemail/services/SentEmailService.scala]: Yep. I removed most of them but seem to have missed this one. Gone now

**Pull request merged**